mac/Keyman4MacIM/Keyman4MacIM/TextApiCompliance.m

```diff
@@ -184,19 +192,85 @@ -(void) checkComplianceAfterInsert:(NSString *)insertedText deleted:(NSString *)
   }
   
-  NSString *message = [NSString stringWithFormat:@"checkComplianceAfterInsert, self.hasWorkingSelectionApi = %@ for app %@", self.hasCompliantSelectionApi?@"YES":@"NO", self.clientApplicationId];
+  // only validate ability to read if we know we have ability to get selection
+  if (self.hasCompliantSelectionApi) {
+    [self validateCanReadInsertedText:insertedText location:selectionRange.location];
+  }
+  
+  NSString *message = [NSString stringWithFormat:@"checkComplianceAfterInsert, self.canReadText = %@, self.hasWorkingSelectionApi = %@ for app %@", self.canReadText?@"YES":@"NO", self.hasCompliantSelectionApi?@"YES":@"NO", self.clientApplicationId];
   os_log_debug([KMLogs complianceLog], "%{public}@", message);
   [KMSentryHelper addDebugBreadCrumb:@"compliance" message:message];
   [KMSentryHelper addApiComplianceTag:self.shortSentryTagDescription];
 }
 
+/**
+ * Confirm that we can get the string that ends at the current location,
+ * and the end of the string matches the string that  was just inserted
+ */
+- (void)validateCanReadInsertedText:(NSString *)insertedText location: (NSUInteger)newLocation {
+
+  NSString *context = [self getContext:newLocation];
+  
+  if (context && context.length > 0) {
+    if ([context hasSuffix:insertedText]) {
+      self.hasCompliantAttributedSubstringApi = YES;
+      os_log_debug([KMLogs complianceLog], "  validateCanReadInsertedText passed, context: '%{public}@', insertedText '%{public}@'", context, insertedText);
+    } else {
+      self.hasCompliantAttributedSubstringApi = NO;
+      os_log_debug([KMLogs complianceLog], "  validateCanReadInsertedText failed, context: '%{public}@' does not have suffix of insertedText '%{public}@'", context, insertedText);
+    }
+  } else {
+    self.hasCompliantAttributedSubstringApi = NO;
+    os_log_debug([KMLogs complianceLog], "  validateCanReadInsertedText failed, attributedSubstringFromRange returned nil or empty string");
+  }
+
+  self.complianceUncertain = NO;
+  self.didValidateAttributedSubstringApi = YES;
+}
+
+/**
+ * Get the string that ends with the current location using `attributedSubstringFromRange`
+ */
+- (NSString *)getContext: (NSUInteger)newLocation {
+  NSAttributedString *attributedString = nil;
+  NSRange contextRange = [self calculateContextRange: newLocation];
+  
+  attributedString = [self.client attributedSubstringFromRange:contextRange];
+  os_log_debug([KMLogs complianceLog], "getContext substring = '%{public}@', length %lu character range: %{public}@: client: %p", attributedString.string, contextRange.length, NSStringFromRange(contextRange), self.client);
+  
+  if (attributedString) {
+    return attributedString.string;
+  } else {
+    return nil;
+  }
+}
+
+/**
+ * Calculate the range which bounds the context that we want to read, up to 80 characters.
+ */
+- (NSRange) calculateContextRange: (NSUInteger)newLocation {
+  const NSInteger kMaxContextLength = 80;
```

Comment:
Ideally this constant is in a header rather than buried here. See for example MAXCONTEXT in kmx_context.h

Reply:
I'll add a TODO for refactoring this. I do not want this constant on a wide scope as it is just for the compliance check. By even existing here, it is almost completely duplicating code in KMInputMethodEventHandler where we do the real fetch of the context for key processing. The compliance check should not be the owner of our default context size.

I realized when making this change that there are multiple places where the code could be improved with a wrapper around the text input client. There are three APIs provided, but they don't give us exactly what we want. Instead we should have methods that wrap those which can be used during both key processing and compliance checks.

Refactoring that will have a bigger impact, so I want to avoid it during beta.